### ISSUE TITLE
Ability to disable/hide options that have already been selected

### ISSUE THREAD

**Issue by bensampaio** [select2/select2#4105]
Is it possible to disable options in the dropdown in some way? In a multi select I would like to be able to hide the already selected options from the dropdown. I tried to use `display: none;` but when I navigate through the options with the arrow keys the hidden options are still highlighted. I'm also trying to hide the new tag option when it is a duplicate of an already available option, but since it is the first element in the dropdown if the user presses enter this will be the selected one even if not visible...

If not possible to disable is there an event being triggered on highlight? This way I could mark the options as disabled and trigger the highlight event in the next one if the current one is disabled / hidden.

**Comment by marcofugaro** [comment]
:+1:

**Comment by JamoCA** [comment]
I have a use case for this too.  I was attempting to stylize the "disabled" options so it was visually obvious which elements were flagged as disabled without having to mouseover any scrutinize the cursors hover state or add text to the end of the displayed value.  (The Select2 demo for disabled options requires manually adding text "(disabled)" to the end of the label to indicate the option is disabled.)  There's no special disabled CSS class to the element, only `aria-disabled="true"`.

After a little research, I've found that I could apply styles to disabled options by using the following CSS aria selector. (NOTE: I'm not sure if this will work in all browsers. I tested it using Firefox Dev 46.):

    .select2-results__option[aria-disabled="true"] {color:#f00 !important;}

**Comment by 719media** [comment]
Hello
I came across this same solution earlier today, however there is an issue with navigation. When using arrow keys to navigate, the hidden options still "take up" an entry. So perhaps there is an improvement that could be made in this regard such that arrow navigation could skip over hidden options?

This could be done by changing the results:next trigger selector line from

``
var $options = self.$results.find('[aria-selected]');
``

``
var $options = self.$results.find('[aria-selected=false]');
``

but, i am not very familiar with this library to know of any drawbacks to this approach

**Comment by alexweissman** [comment]
There is a good reason that you wouldn't want to remove a selected item from the list completely.  See Kevin's explanation in #1538:

> Right now the only way to unselect arbitrary (other than the last) options using the keyboard is to type CTRL + SPACE when the result is highlighted. This is to command for toggling a result, so it will select or unselect the highlighted result depending on what state it was previously in. Screen readers should read out the state of the option (whether it is selected or not) as they move through the results list, which is usually important.

> Without an alternate way to unselect arbitrary options using the keyboard, the UX problems caused by that far outweigh the UX annoyances that you currently have by showing the selected results.

This should probably be documented since it is such a common question.

**Comment by multiwebinc** [comment]
If this is an accessibility choice, would it not make more sense for screen readers to simply use the original `<select>` item instead of select2? What is the benefit of screen readers using select2 at all? (Please note, I'm not trying to talk negative about select2, but I genuinely don't know the answer. You could also simplify the code to not use any `aria` attributes).

Anyway, here's a way that allows you to hide the elements from browsers and keep them "visible" for screen readers:

```
.select2-results__option[aria-selected="true"] {
    position: absolute !important;
    clip: rect(1px 1px 1px 1px); /* In case you want to support IE6, IE7 */
    clip: rect(1px, 1px, 1px, 1px);
}
```

And then to fix the arrow keys, as @719media more or less said above, in the `results:previous` and `results:next` functions, simply do:

    var $options = self.$results.find('[aria-selected=false]');

This *should* be the best of both worlds. Then you could include some sort of [configuration option](https://select2.org/options) to accomplish this programatically. I'm assuming you would want to use the javascript to conditionally add some sort of `data` attribute to the selected options (in addition to `aria`), then the CSS would also match `data` instead of `aria`.

**Comment by alexweissman** [comment]
I don't think it's just about screen readers, though.  There are lots of contexts where you might see the Select2 control, but only be able to interact with it via the keyboard.

You're right though, this shouldn't be blocking on a configurable option to hide selected items.  Perhaps the ideal solution would be to make the actual selection pills navigable, so that they could be unselected directly via keyboard.  Then options could be safely hidden without the problem of items being unremovable in keyboard-only contexts.

**Comment by multiwebinc** [comment]
> Perhaps the ideal solution would be to make the actual selection pills navigable, so that they could be unselected directly via keyboard

That's how StackOverflow handles it for the question tagging. Pushing the left arrow when on the leftmost character changes the current tag back into a pill and makes the tag to the left become editable. The opposite for pushing the right arrow when there is a pill after the current tag being edited.

**Comment by alexweissman** [comment]
Ah, I was talking about tabbing through the `x` buttons to be able to actually clear entire selections (with or without tagging enabled).

**Comment by Grezvany13** [comment]
Already commented at #4991 , but since it's closed I'll add it here as well.

I would suggest to be able to modify the original ```select``` and then properly display (after re-initialising).

I already saw several issues related to the fact that Select2 doesn't care about CSS styling (```display:none```), and I even noticed that it doesn't even care about the ```hidden``` attribute.

CodePen: https://codepen.io/anon/pen/JrJKBq

As you can see the default handling (at least in Chrome) will hide the options which have either ```display:none``` in the CSS or have the ```hidden``` attribute. At least it follows the ```disabled``` attribute.

**Comment by Darkov3** [comment]
For anyone else experiencing difficulties regarding this issue, as I did:

Finally, I reached the following solution(until the select2 lib supports this by itself).
I didn't need multiple selects however and I have not tested it in all browsers, so you may want to investigate these things if they are of concern to you.

First, to hide the selected value, I used the already mentioned CSS solution:
```css
.select2-dropdown .select2-results__option[aria-selected=true] {
    display: none;
}
```

Next, to fix the arrow keys issue, I set up some event handling with JQ:
```javascript
$selectElement.data('select2').on('results:next', function (event) {
    // find the currently highlighted element:
    var $currentSelection = $selectElement.data('select2').$dropdown.find('.select2-results__option--highlighted');
    // findout if its the currently selected element:
    if ($currentSelection.is('[aria-selected="true"]')) {
        // if it is, we trigger another key event, to skip this element
        // the following if is required to avoid recursion and to achieve default select2 behaviour
        if ($currentSelection.is(':last-child')) {
            // if its the last child, while pressing the down arrow key, we need to go back to the previous one, instead of going to the next one
            $selectElement.data('select2').trigger('results:previous');
        } else {
            // if its not the last child, while pressing the down arrow key, just go forward to the next element and skip this one
            $selectElement.data('select2').trigger('results:next');
        }
    }
});

// same as the first event, but in reverse:
$selectElement.data('select2').on('results:previous', function (event) {
    var $currentSelection = $selectElement.data('select2').$dropdown.find('.select2-results__option--highlighted');
    if ($currentSelection.is('[aria-selected="true"]')) { 
        if ($currentSelection.is(':first-child')) {
            $selectElement.data('select2').trigger('results:next');
        } else {
            $selectElement.data('select2').trigger('results:previous');
        }
    }
});
```

And finally, my select dropdown has round border edges, so I needed to add border-radius only to the last element in the select list. But if the last element is the selected one, it will be invisible. After a few hours of trying to fix this using CSS, I figured that is impossible with CSS(maybe I am wrong). So I ended up using JQ to add a special CSS selector to the element that is the last **visible** select option element:

```javascript
// this is binded to the highlighted elements only
$selectElement.data('select2').$dropdown.on('mouseenter', '.select2-results__option--highlighted', function () {
    if ($(this).is(':last-child')) {
        // if the last child is the current element, then give the css class to the prev element, otherwise give it to this element
        if ($(this).is('[aria-selected="true"]')) {
            if (!$(this).prev().hasClass('select2-real-last-child-fix')) {
                // remove any current elements with this css, can happen under special circumstances
                self.$statusSelect.data('select2').$dropdown.find('.select2-results__option').removeClass('select2-real-last-child-fix');
                $(this).prev().addClass('select2-real-last-child-fix');
            }
        } else {
            // same as the other if, but applied to this element, instead of the prev element
            if (!$(this).hasClass('select2-real-last-child-fix')) {
                self.$statusSelect.data('select2').$dropdown.find('.select2-results__option').removeClass('select2-real-last-child-fix');
                $(this).addClass('select2-real-last-child-fix');
            }
        }
    }
});
```

**Comment by legshooter** [comment]
The `templateResult` callback option seems to work great for me:
```js
templateResult: function formatState (state) {
    return state.selected ? null : state.text;
},
```

**Comment by mkasprz** [comment]
It is possible to use `sorter` function to remove such elements. It's quite sad though, that the "No data" message is checked to be displayed before sorting, not after it. That's why in the following snippet I just close the dropdown if it's about to be empty.
```typescript
$(this.dropdown).select2({sorter: (data) => {
   let results = data.filter(item => !item.selected);
   if (data.length && !results.length) {          
       $(this.dropdown).select2('close');
   }
   return results;
}
```

**Comment by pedrofurtado** [comment]
We don't have immediate plans to provide this. We are focused to fix some major UI bugs (that are majority of issues and PR's). But if you open a PR with unit tests, I will be glad to review and approve if everything is ok :+1:

**Comment by henray** [comment]
+1

**Comment by theAlinux** [comment]
```
$("body").on('click', 'button.edit', function () {
        $('select.extra_future').val(null).trigger('change');
        var features = $(this).data('data');
        $.each(features, (ind, val) => {

            $('select.extra_future option[value="' + val + '"]').prop('selected', true).trigger('change');

            $('select.extra_future option[value="' + val + '"]').closest('optgroup').find('option').attr('disabled', true).trigger('change');
            $('select.extra_future option[value="' + val + '"]').attr('disabled', false).trigger('change');

            $('.extra_future').select2({
                placeholder: "SELECT",
                allowClear: true
            });
            
        })

    });
```

**Comment by schaerli** [comment]
+1

**Comment by laflier** [comment]
Here's a workaround for cascading dropdowns using Select2, without Ajax.

https://codepen.io/lflier/pen/xxWXwpg

**Comment by abasina09-333** [comment]
$(document).on('mouseenter', '.select2-results__option', function () {
  $(this).css('background-color', '#97c121');
  $(this).css('color', 'white');
});

// When the mouse leaves an option, remove the custom background color
$(document).on('mouseleave', '.select2-results__option', function () {
  $(this).removeClass('select2-results__option--highlighted');
  $(this).css('background-color', '');
  $(this).css('color', '');
});
this works for me
